[Static runtime] refactor static runtime to expose useful liveness methods

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63873
* #64350
* #64349
* #64348
* #64347
* __->__ #62995

Differential Revision: [D30769101](https://our.internmc.facebook.com/intern/diff/D30769101)

makslevental committed:
refactor static runtime to expose useful liveness methods

[ghstack-poisoned]

Commit:
Update on "refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

desertfire reviewed (approved):
LGTM. @hlu1, any specific Static Runtime test needs to be run here?

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

makslevental commented:
> LGTM. @hlu1, any specific Static Runtime test needs to be run here?

@hlu1 I have refactored this so that it is easier to review now. What do you think?

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -121,6 +121,7 @@ bool mayContainAlias(
   }
   return db.mayContainAlias(as, bs);
 }
+}
```
Where does the extra right bracket come from?

makslevental replied:
opening and closing the anonymous namespace - i did this so that you could more easily see the diff. after you approve i will move these two functions lower out of the anonymous namespace.

hlu1 commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -114,7 +114,7 @@ class TORCH_API StaticModule {
   explicit StaticModule(
       std::pair<std::shared_ptr<torch::jit::Graph>, std::shared_ptr<Module>>
-          graph_and_module,
+      graph_and_module,
```
Could you clang-format the files?

makslevental replied:
clang-format actually added this indent. i'll remove by hand

hlu1 commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -454,5 +456,57 @@ class TORCH_API ProcessedNode {
+
+using LiveRangesMap = std::unordered_map<const Value*, LiveRange>;
+
+TORCH_API std::unordered_set<const Value*> GetAlwaysAliveValues(
```
Return type doesn't match the the implementation.

makslevental replied:
ah sorry `FastSet <=> unordered_set` got away from me.

hlu1 commented on torch/csrc/jit/runtime/static/impl.h:
```diff
@@ -454,5 +456,57 @@ class TORCH_API ProcessedNode {
+  return !(lhs == rhs);
+}
+
+using LiveRangesMap = std::unordered_map<const Value*, LiveRange>;
```
`FastMap` is the same as `std::unordered_map` in OSS
```
using LiveRangesMap = FastMap<const Value*, LiveRange>;
```

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -173,15 +171,22 @@ LivenessMap GetLivenessMap(
   // map Values to its creation order in graph (Note: only traverse top-level
   // nodes such that nodes under control-flows are represented by top-level
   // block nodes)
+  int idx = 1;
```
Use `size_t` directly? Why are you starting from 1 not 0? Does it matter?

makslevental replied:
0/1 doesn't matter but what do you mean by "Use `size_t` directly? just

`size_t idx = 0` ? if so then sure.

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -173,15 +171,22 @@ LivenessMap GetLivenessMap(
+  int idx = 1;
   std::vector<const Value*> values_in_creation_order;
   FastMap<const Value*, size_t> values_to_idx_in_creation_order;
+  std::unordered_map<const Node*, size_t> nodes_to_idx_in_topo_order;
```
Use `FastMap` here.

hlu1 commented on torch/csrc/jit/runtime/static/impl.cpp:
```diff
@@ -173,15 +171,22 @@ LivenessMap GetLivenessMap(
     }
   }
 
+  std::unordered_map<const Value*, size_t> value_creation_topo_idx;
+  std::unordered_map<const Value*, std::unordered_set<size_t>>
+      value_use_topo_idxs;
```
Ditto. Use `FastSet` for `std::unordered_set`

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"




[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769101).

Commit:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"


Differential Revision: [D30769101](https://our.internmc.facebook.com/intern/diff/D30769101)

[ghstack-poisoned]

makslevental committed:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"


Differential Revision: [D30769101](https://our.internmc.facebook.com/intern/diff/D30769101)

[ghstack-poisoned]

makslevental committed:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"


Differential Revision: [D30769101](https://our.internmc.facebook.com/intern/diff/D30769101)

[ghstack-poisoned]

makslevental committed:
Update on "[Static runtime] refactor static runtime to expose useful liveness methods"


Differential Revision: [D30769101](https://our.internmc.facebook.com/intern/diff/D30769101)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769101).
